wild_lib/src/elf.rs

```diff
@@ -121,7 +123,40 @@ impl<'data> File<'data> {
     }
 
     pub(crate) fn section_data(&self, section: &SectionHeader) -> Result<&'data [u8]> {
-        Ok(section.data(LittleEndian, self.data)?)
+        let data = section.data(LittleEndian, self.data)?;
+
+        if let Some((compression, offset, _)) = section.compression(LittleEndian, self.data)? {
+            let format = match compression.ch_type.get(LittleEndian) {
+                object::elf::ELFCOMPRESS_ZLIB => object::CompressionFormat::Zlib,
+                object::elf::ELFCOMPRESS_ZSTD => object::CompressionFormat::Zstandard,
+                c => bail!("Unsupported compression format: {}", c),
+            };
+
+            let decompressed = CompressedData {
+                format,
+                data: &data[(offset as usize)..],
+                uncompressed_size: compression.ch_size.get(LittleEndian),
+            }
+            .decompress()?;
+            // TODO: find a proper lifetime for the uncompressed data!!!
+            Ok(decompressed.into_owned().leak())
```

Comment:
@davidlattimore: The following code should support getting data for a compressed section. The main question is how to cache the results and how to properly deallocate the uncompressed vector. Any suggestions?

Reply:
Is it only debug sections that use compression?

Do we need access to the section data during layout, or only when we're writing the output?

For executable sections, we currently look at the instruction bytes during layout to work out if we can apply certain optimisations. If we can apply those optimisations, then that affects whether we need to allocate space in the GOT, PLT etc. I'd hope that we don't need to do anything like that for debug info.

If we don't, then it might be possible to decompress directly into the output. That would be ideal for efficiency, since there's less copying and less heap allocation.

Reply:
Although I just looked at the code that calls `Relaxation::new` in `layout.rs` and it does it unconditionally. I suspect that's (a) potentially wrong and (b) potentially inefficient. So that bit of code can probably be changed to only call it if the section has the execute bit set. That would eliminate one call to `section_data`.

I see that I'm also storing a slice of section data on `Section`. From a quick look that's mostly not used until we write. Possibly something there can be refactored to delay that call to `section_data`.

Reply:
> Is it only debug sections that use compression?

Yes. Some time ago, `maskray` suggested an LLD option for compression of any section, but he abandoned the effort: https://maskray.me/blog/2023-07-07-compressed-arbitrary-sections.

> Do we need access to the section data during layout, or only when we're writing the output?

Yes. As shown [here](https://github.com/davidlattimore/wild/issues/37#issuecomment-2310198009), debug info sections utilize relocations quite a lot and thus the section should be properly loaded and the relocations resolved (AFAIK debug info should not use any dynamic resolutions).

Reply:
> Yes. As shown [here](https://github.com/davidlattimore/wild/issues/37#issuecomment-2310198009), debug info sections utilize relocations quite a lot and thus the section should be properly loaded and the relocations resolved (AFAIK debug info should not use any dynamic resolutions).

For relocations in executable code, we process the twice, once during layout to (a) do a graph traversal to work out what is referenced and (b) work out if we need to allocate any space for a runtime relocation - e.g. for a GOT entry or similar. Then during file writing, we process the relocations again to actually apply them. It sounds to me like perhaps only the second lot of processing is needed for debug sections. i.e. the relocations need to be applied, but we don't need to have done anything with them during graph traversal (`layout.rs`).

Reply:
Yes, I understand it the same. That sounds promising!

Reply:
Regarding the current approach: `decompressed.into_owned().leak()`: is it something you can live for now or do you have any recommendations on what to do instead?

Reply:
I guess it only runs when we encounter compressed sections, which we currently don't support at all, so it's not regressing anything that we currently support. In that regard, I'd be OK with leaking in the short-term - assuming we want to make the refactorings required to not need it afterwards